File: src/vscode_py/vscode_py/incrementer_node.py
import rclpy
import rclpy.node

from rclpy.executors import ExternalShutdownException

import std_msgs.msg


class Incrementer:
    def __init__(self, node: rclpy.node.Node):
        self.node = node

        self.node.get_logger().info("Initailizing incrementer node")

        # Create publisher
        self.publisher = self.node.create_publisher(
            msg_type=std_msgs.msg.Int32, topic="incremented_number", qos_profile=10
        )

        # Create subscriber
        self.subscriber = self.node.create_subscription(
            msg_type=std_msgs.msg.Int32,
            topic="number",
            callback=self.callback,
            qos_profile=10,
        )

    def callback(self, msg: std_msgs.msg.Int32):
        self.node.get_logger().info(f"In callback with number: {msg.data}")

        number = msg.data
        incremented_number = number + 1

        # Publish incremented number
        self.publisher.publish(std_msgs.msg.Int32(data=incremented_number))


def main(args=None):
    # Init rclpy
    rclpy.init(args=args)

    # Create node
    node = rclpy.create_node("incrementer")
    _ = Incrementer(node)

    # Spin node until destroyed
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == "__main__":
    main()

File: src/vscode_cpp/src/incrementer_node.cpp
// Copyright 2023 Samuel Lindgren

#include "incrementer_node.hpp"

IncrementerNode::IncrementerNode() : Node("incrementer") {
  RCLCPP_INFO(this->get_logger(), "Incrementer node started");

  step_size_ = 1; // Default step size

  // Allow step size to be set from the launch file
  declare_parameter("step_size", step_size_);
  get_parameter("step_size", step_size_);

  incremented_number_pub_ =
      this->create_publisher<std_msgs::msg::Int32>("incremented_number", 10);

  auto number_callback = [this](std_msgs::msg::Int32::SharedPtr msg) {
    auto incremented_number_msg = std::make_unique<std_msgs::msg::Int32>();
    int number = msg->data;
    int incremented_number = number + step_size_;
    incremented_number_msg->data = incremented_number;
    incremented_number_pub_->publish(std::move(incremented_number_msg));
  };

  number_sub_ = this->create_subscription<std_msgs::msg::Int32>(
      "number", 10, number_callback);
}

IncrementerNode::~IncrementerNode() {}

void IncrementerNode::set_step_size(int step_size) { step_size_ = step_size; }

int main(int argc, char *argv[]) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<IncrementerNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}